### PR TITLE
Report generation

### PR DESCRIPTION
**Pull request by sylvesteraswin**
1. Setup `webpack` with `preact` to create the app logic for the report.
1. Setup `ejs` to create the template and pass `data` to create the report.
1. Setup hooks to get the `duplicates` while printing.
1. Format `data` to extract only the necessary fields to pass to the report.

@boopathi @vigneshshanmugam 

### Pending Tasks
- [ ]  View switcher: Ability to switch between duplicates by count and duplicates by lines.
- [x]  Build Accordion to show only the 1st duplicate. Other duplicates should be inside an accordion so uses can open and see by demand.
- [x]  Pretty print only block whats necessary beyond a pre-configured threshold.
- [ ]  Create the logic to compute number of duplicates in a single build file.
- [ ]  Support huge build files.

### CODE REVIEW

File: client/.babelrc [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+      {
+        "targets": {
+          "browsers": ["> 5%", "last 2 versions"]
+        }
+      }
+    ]
+  ],
+  "plugins": [
+    ["transform-react-jsx", { "pragma": "h" }],
+    "transform-class-properties"
```
Comment (boopathi) [inline client/.babelrc]: don't use this

File: client/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    ]
+  },
+  plugins: (plugins => {
+    if (isDev) {
+      // plugins.push([new webpack.NamedModulesPlugin()]);
+      return plugins;
+    }
+    plugins.push(new BabiliPlugin());
+    return plugins;
+  })([])
```
Comment (boopathi) [inline client/webpack.config.js]: no IIFE

File: package.json [hunk 3]
```diff
@@ -20,12 +30,41 @@
...
+    "mkdirp": "0.5.1",
+    "normalize.css": "7.0.0",
+    "postcss": "6.0.8",
+    "postcss-apply": "0.8.0",
+    "postcss-cssnext": "3.0.2",
+    "postcss-import": "10.0.0",
+    "postcss-import-url": "2.2.0",
+    "postcss-loader": "2.0.6",
+    "postcss-mixins": "6.0.1",
+    "postcss-nesting": "4.0.1",
```
Comment (boopathi) [inline package.json]: postcss here is an overkill .. remove all these plugins and simply use css-modules.

File: client/index.jsx [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import Reporter from "./components/Reporter";
+
+import styles from "./style/index";
+
+window.addEventListener(
+  "load",
+  () => {
+    renderReport(window.duplicates);
+  },
+  false
```
Comment (vigneshshanmugam) [inline client/index.jsx]: not required.

Comment (vigneshshanmugam) [inline client/index.jsx]: is this required?

File: lib/cli.js [hunk 5]
```diff
@@ -114,8 +115,16 @@ if (program.filelist) {
 }
 
 function print(cpa) {
-  const report = new Reporter(cpa.findOptimalDuplicates());
+  const duplicates = cpa.findOptimalDuplicates();
+  const report = new Reporter(duplicates);
   process.stdout.write(report.toString({ colors: program.colors }));
+
+  generateHTMLReport(duplicates);
```
Comment (vigneshshanmugam) [inline lib/cli.js]: add a new flag to cli `--report` and call this func

File: lib/cli.js [hunk 6]
```diff
@@ -7,6 +7,7 @@ const program = require("commander");
 
 const CPA = require("./cpa");
 const Reporter = require("./reporter");
+const HTMLReporter = require("./htmlReporter");
```
Comment (vigneshshanmugam) [inline lib/cli.js]: use kebab case for filename, html-reporter

File: client/components/Sidebar/Sidebar.jsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    }
+    console.log(!list[0] || !list[0].matchingCode);
+    if (!list[0] || !list[0].matchingCode) {
+      return null;
+    }
+    console.log(list[0].matchingCode.length);
+    return list[0].matchingCode.length;
+  };
+  return (
+    <div className={BASE_CLASS}>
```
Comment (boopathi) [inline client/components/Sidebar/Sidebar.jsx]: use css modules

File: client/webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+module.exports = {
+  context: __dirname,
+  entry: "./index",
+  output: {
+    path: path.join(__dirname, "..", "public"),
+    filename: "reporter.js",
+    publicPath: "/"
+  },
+  resolve: {
+    extensions: [".js", ".jsx", ".css"]
```
Comment (boopathi) [inline client/webpack.config.js]: don't use `css` in resolve. Using `css` while importing is explicit and clear.

File: package.json [hunk 9]
```diff
@@ -20,12 +30,41 @@
   },
   "dependencies": {
     "babel-types": "^7.0.0-alpha.15",
-    "babylon": "^7.0.0-alpha.16",
+    "babylon": "^v7.0.0-beta.18",
```
Comment (boopathi) [inline package.json]: remove `v`

File: package.json (hunk 3 above)
Comment (sylvesteraswin) [reply package.json]: 💃  I will see how to remove it 🗡

File: package.json [hunk 10]
```diff
@@ -10,9 +10,18 @@
     "test": "jest",
     "build": "NODE_ENV=production webpack --config client/webpack.config.js"
   },
-  "keywords": ["code", "repetition", "pattern", "duplication", "javascript"],
```
Comment (vigneshshanmugam) [inline package.json]: disable prettier from dealing with json files.

File: client/components/PrettyPrint/PrettyPrint.jsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import "prismjs/plugins/line-highlight/prism-line-highlight";
+
+import "./prettyprint.css";
+
+const BASE_CLASS = "pretty-print";
+
+export default class PrettyPrint extends Component {
+  constructor(...args) {
+    super(...args);
+    this.highlight = this.highlight.bind(this);
```
Comment (boopathi) [inline client/components/PrettyPrint/PrettyPrint.jsx]: this is not required

File: client/components/Sidebar/Sidebar.jsx [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+import cx from "classnames";
+
+import "./sidebar.css";
+
+const BASE_CLASS = "sidebar";
+
+const Sidebar = ({ data, handleItemChange, activeIndex }) => {
+  // Get the length of the matching function
+  const getFirstFunctionMatchLength = list => {
+    if (!list.length) {
```
Comment (boopathi) [inline client/components/Sidebar/Sidebar.jsx]: Use `list.length === 0`

File: client/components/Sidebar/Sidebar.jsx [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+      return null;
+    }
+    if (!list[0] || !list[0].matchingCode) {
+      return null;
+    }
+    return list[0].matchingCode.length;
+  };
+  if (!data || !data.length) {
```
Comment (boopathi) [inline client/components/Sidebar/Sidebar.jsx]: use `data.length < 1` or `data.length === 0`

File: client/index.jsx [hunk 14]
```diff
@@ -0,0 +1,18 @@
+import { h, render } from "preact";
+
+import Reporter from "./components/Reporter";
+
+import "./style/index.css";
+
+window.addEventListener("load", () => {
```
Comment (boopathi) [inline client/index.jsx]: move script in template to end of body and avoid onload.

File: lib/cli.js [hunk 15]
```diff
@@ -29,6 +30,7 @@ program
     100
   )
   .option("-C, --no-colors", "Disable colors in output")
+  .option("-R, --report", "Generate HTML report")
```
Comment (boopathi) [inline lib/cli.js]: Use: `-R html` | `--report html` instead of boolean `-R` ??

File: client/webpack.config.js [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+                      path: path.join(__dirname, "postcss.config.js")
+                    }
+                  }
+                }
+              ]
+            })
+      }
+    ]
+  },
+  plugins: [].concat(
```
Comment (vigneshshanmugam) [inline client/webpack.config.js]: add ModuleConcatenationPlugin as well to enable scope hoisiting

File: lib/cli.js [hunk 17]
```diff
@@ -60,7 +62,10 @@ if (program.filelist) {
   rl.on("close", () => {
     clearBuffer();
     try {
-      print(cpa);
+      const duplicates = print(cpa);
```
Comment (boopathi) [inline lib/cli.js]: why are you depending on the output of print function ? This seems wrong.

File: lib/html-reporter.js [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+  return fs.readFileSync(`${projectRoot}/public/${filename}`, "utf-8");
+}
+
+class HTMLReporter {
+  constructor(duplicates, opts) {
+    this.duplicates = duplicates;
+
+    this.opts = Object.assign({}, defaultOptions, opts);
+
+    this.renderFile = this.renderFile.bind(this);
```
Comment (boopathi) [inline lib/html-reporter.js]: this is not required

File: lib/html-reporter.js [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    this.printError = this.printError.bind(this);
```
Comment (boopathi) [inline lib/html-reporter.js]: this is not required

File: package.json [hunk 20]
```diff
@@ -20,12 +30,42 @@
...
   "devDependencies": {
-    "jest": "^20.0.4"
+    "babel-core": "6.25.0",
+    "babel-loader": "7.1.1",
+    "babel-plugin-transform-react-jsx": "6.24.1",
+    "babel-preset-env": "1.6.0",
+    "babili-webpack-plugin": "0.1.2",
+    "classnames": "2.2.5",
+    "css-loader": "0.28.4",
+    "exports-loader": "0.6.4",
```
Comment (boopathi) [inline package.json]: why is this required ?

File: package.json [hunk 21]
```diff
@@ -20,12 +30,42 @@
...
+    "extract-text-webpack-plugin": "3.0.0",
+    "jest": "^20.0.4",
+    "mkdirp": "0.5.1",
```
Comment (boopathi) [inline package.json]: why is this required? isn't `ejs` taking care of it ?

File: client/components/Header/header.css [hunk 22]
```diff
@@ -0,0 +1,32 @@
+@import '../../style/colors';
+@import '../../style/variables';
+@import '../../style/mixins';
+
+:root {
```
Comment (boopathi) [inline client/components/Header/header.css]: why always declare in `root`?

File: client/components/Content/content.css [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+  overflow-y: auto;
+  overflow-x: hidden;
+  flex-grow: 1;
+  z-index: 1;
+  -webkit-overflow-scrolling: touch;
+  margin-left: var(--sidebar-width);
+  padding-left: var(--content-padding-left);
+  padding-top: var(--content-padding-left);
+
+  & .print-wrapper {
```
Comment (boopathi) [inline client/components/Content/content.css]: doesn't need to be nested for 3 classes. Move to css-modules and remove postcss-nested.

File: client/components/Content/Content.jsx [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+import cx from "classnames";
+
+import PrettyPrint from "../PrettyPrint";
+
+import "./content.css";
+
+const BASE_CLASS = "content";
+
+const Content = ({ data }) => {
+  if (!data || !data.length) {
```
Comment (boopathi) [inline client/components/Content/Content.jsx]: `data.length === 0`

File: client/index.jsx [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+  renderReport(window.duplicates);
+});
+
+let app;
+function renderReport({ data }) {
+  app = render(
+    <Reporter data={data} />,
+    document.getElementById("reporterApp"),
+    app
```
Comment (boopathi) [inline client/index.jsx]: why is this required?

File: package.json (hunk 21 above)
Comment (sylvesteraswin) [reply package.json]: `ejs` does not create a folder if it does not exist, it only creates or replaces a file. Since we will be creating a `report` in the desired folder specified by the user, we will have to make sure the folder exists.

File: package.json (hunk 20 above)
Comment (sylvesteraswin) [reply package.json]: This is not required, will remove it.

File: client/components/Header/header.css (hunk 22 above)
Comment (sylvesteraswin) [reply client/components/Header/header.css]: > By declaring a variable on the :root pseudo-class, a CSS author can halt some instances of repetition by using the variable.

As far as I know or seen `css variables` are always declared inside `:root`.

File: lib/cli.js (hunk 17 above)
Comment (sylvesteraswin) [reply lib/cli.js]: Like I told you `findOptimalDuplicates` function in `cpa` object always returns a `[]` if I call it from `generateHTMLReport`(calling it second time). You need to reset the memory in that function so I can use it correctly.

File: client/webpack.config.js (hunk 16 above)
Comment (sylvesteraswin) [reply client/webpack.config.js]: done.

File: lib/cli.js (hunk 15 above)
Comment (sylvesteraswin) [reply lib/cli.js]: Done. But how do I catch an `error` when user passes a null value.

File: client/index.jsx (hunk 14 above)
Comment (sylvesteraswin) [reply client/index.jsx]: I wanted the app to execute only during the `onload` function. Is there anything wrong in having this, since this is a standalone file.

File: client/components/Sidebar/Sidebar.jsx (hunk 12 above)
Comment (sylvesteraswin) [reply client/components/Sidebar/Sidebar.jsx]: Is there any reason to check this explicitly?

File: client/components/Sidebar/Sidebar.jsx (hunk 13 above)
Comment (sylvesteraswin) [reply client/components/Sidebar/Sidebar.jsx]: Is there any reason to check this explicitly?

File: client/components/Content/Content.jsx (hunk 24 above)
Comment (sylvesteraswin) [reply client/components/Content/Content.jsx]: Is there any reason to check this explicitly?

File: client/components/Content/content.css (hunk 23 above)
Comment (sylvesteraswin) [reply client/components/Content/content.css]: I will move this today.

File: client/index.jsx (hunk 25 above)
Comment (sylvesteraswin) [reply client/index.jsx]: Third argument, is the root node to replace, otherwise it appends.

File: client/components/PrettyPrint/PrettyPrint.jsx (hunk 11 above)
Comment (sylvesteraswin) [reply client/components/PrettyPrint/PrettyPrint.jsx]: If I don't do this it not have access to `props` and `state`. It does not auto-bind.

File: client/components/Sidebar/Sidebar.jsx (hunk 12 above)
Comment (boopathi) [reply client/components/Sidebar/Sidebar.jsx]: It's explicit and clear.

Would be micro-optimisation: but explicits are performant than implicit boolean typecasting.

File: client/components/Content/Content.jsx [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+import styles from "./content.css";
+
+const Content = ({ data }) => {
+  if (!data || !data.length) {
+    return;
+  }
+  return (
+    <div className={styles.content}>
+      <div className={styles.printWrapper}>
+        {(() => {
```
Comment (boopathi) [inline client/components/Content/Content.jsx]: why is this extra IIFE required ?

```js
{
  data.map(d => ...);
}
```
?

File: client/components/Content/index.js [hunk 27]
```diff
@@ -0,0 +1,2 @@
+import Content from "./Content";
+export default Content;
```
Comment (boopathi) [inline client/components/Content/index.js]: Do components need their own directories?

I think it's simpler to just use 

```
components/
  - Content.js
  - Content.css
  - Header.js
  - Header.css
```

Instead of creating a separate directory per component, which seems unnecessary, just keep all the files flat. And these dummy `index.js` files won't be needed. Also when the component exports something other than `default`, it has to be added in `index.js` as well.

File: client/components/PrettyPrint/PrettyPrint.jsx (hunk 11 above)
Comment (boopathi) [reply client/components/PrettyPrint/PrettyPrint.jsx]: I see the only place of invoking `highlight` is in `componentDidMount` and it is invoked as `this.highlight()`, and `componentDidMount` is already auto-bound to `this`. So invoking it as 

```js
const x = this.highlight;
x();
```

has no this bound to `highlight`. But invoking it as -

```js
this.highlight();
```

has this binding in highlight. So it should work without this binding.

File: client/index.jsx (hunk 14 above)
Comment (boopathi) [reply client/index.jsx]: https://stackoverflow.com/a/6026730/556124

File: lib/html-reporter.js [hunk 28]
```diff
@@ -0,0 +1,108 @@
+const path = require("path");
+const fs = require("fs");
+const http = require("http");
+const chalk = require("chalk");
+const mkdir = require("mkdirp");
+
+const ejs = require("ejs");
+
+const projectRoot = path.resolve(__dirname, "..");
```
Comment (boopathi) [inline lib/html-reporter.js]: use `path.join`

File: lib/html-reporter.js [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+
+const defaultOptions = {
+  reportFileDir: "public",
+  appFileName: "reporter.js",
+  appStyleName: "reporter.css",
+  reporterFileName: "report.html"
+};
+
+function getAssetsContent(filename) {
+  return fs.readFileSync(`${projectRoot}/public/${filename}`, "utf-8");
```
Comment (boopathi) [inline lib/html-reporter.js]: use `path.join`

File: lib/html-reporter.js [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+      process.stderr.write(
+        chalk.red(
+          chalk.bold("CPA report ::NOT:: created since were 0 duplicates!!")
+        )
+      );
+      return;
+    }
+
+    ejs.renderFile(
+      `${projectRoot}/views/index.ejs`,
```
Comment (boopathi) [inline lib/html-reporter.js]: use `path.join`

File: lib/html-reporter.js [hunk 31]
```diff
@@ -0,0 +1,108 @@
...
+              });
+            })
+            .sort((arr1, arr2) => {
+              return arr2.length - arr1.length;
+            })
+        })
+      },
+      (err, reporterHtml) => {
+        if (err) {
+          this.printError(err);
```
Comment (boopathi) [inline lib/html-reporter.js]: Shouldn't you want to return when it errors ?

File: client/components/Content/Content.jsx (hunk 26 above)
Comment (sylvesteraswin) [reply client/components/Content/Content.jsx]: Ah I used it for someother purpose, but now its not required, will remove it.

File: client/components/PrettyPrint/PrettyPrint.jsx (hunk 11 above)
Comment (sylvesteraswin) [reply client/components/PrettyPrint/PrettyPrint.jsx]: Yes it should work from a `class` perspective. But since this class extends from `preact`'s `Component` it does not work and `this` in this case would be the `super` class we extend from and not the `current` class(I think), so I have to explicitly bind this to the current `this` in the `constructor` for it work properly. 

PS: Had I used `class-properties` it would have auto-bound without me doing this in the `constructor`.

File: client/components/Content/Content.jsx (hunk 26 above)
Comment (sylvesteraswin) [reply client/components/Content/Content.jsx]: Ah I had used this for a different purpose, but now its not required, I will remove it.

File: client/components/Content/index.js (hunk 27 above)
Comment (sylvesteraswin) [reply client/components/Content/index.js]: IMO, having separate folders is easier to understand/view the components which are under the folder. Having a flat structure is confusing .

Once `export * from 'component'` is ready we wont have this issue 💃

Comment (boopathi) [reply client/components/Content/index.js]: But you still need that extra `index.js`. Flat is easier to look up as it's grouped by the names already and the js and css files stand next to each other always and with the benefit that you can use `import x from "./x"` instead of `import x from "./x/x"` which I understand is the purpose of this `index.js`. How is a separate directory easier to understand/view the components?

File: lib/html-reporter.js (hunk 31 above)
Comment (sylvesteraswin) [reply lib/html-reporter.js]: Done. It will print error and `return`.

File: client/components/Content/Content.jsx [hunk 32]
```diff
@@ -5,8 +5,13 @@ import PrettyPrint from "../PrettyPrint";
...
 
-const Content = ({ data }) => {
-  if (!data || !data.length) {
+const Content = ({
+  data,
+  codeLength = 40, // Lines after which entire code shoult not be displayed
+  codeThreshold = 3, // Lines to show before and after the slice
+  baseLine = 1 // Starting line number
+}) => {
+  if (!data || !data.length === 0) {
```
Comment (boopathi) [inline client/components/Content/Content.jsx]: no `!` before `data.length`

File: package.json (hunk 10 above)
Comment (boopathi) [reply package.json]: actually, this is the npm formatting. prettier puts it in a single line.

File: client/index.jsx (hunk 14 above)
Comment (sylvesteraswin) [reply client/index.jsx]: The only problem here is, how do I get the instance of `preact` render function in the `ejs` file.

In this case, everything will be bundle during `webpack` and I don't have to care about resolving the modules.

File: client/components/Content/Content.jsx [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+class Content extends Component {
+  constructor(...args) {
+    super(...args);
+
+    this.state = {
+      activeItem: 0
+    };
+
+    this.handleItemChange = this.handleItemChange.bind(this);
+    this.shallowDiffers = this.shallowDiffers.bind(this);
```
Comment (boopathi) [inline client/components/Content/Content.jsx]: this is not required

File: client/components/Content/Content.jsx [hunk 34]
```diff
@@ -0,0 +1,141 @@
...
+  render(
+    {
+      data,
+      codeLength = 40, // Lines after which entire code should not be displayed
+      codeThreshold = 3, // Lines to show before and after the slice
+      baseLine = 1 // Starting line number
+    },
+    { activeItem }
+  ) {
+    if (!data || !data.length === 0) {
```
Comment (boopathi) [inline client/components/Content/Content.jsx]: Use default assignment instead of `!data` and there should be no `!` before `data.length === 0`

```js
render({ data = {} });
```

File: client/components/Content/Content.jsx [hunk 35]
```diff
@@ -0,0 +1,141 @@
...
+              id,
+              loc: {
+                end: { line: endLine } = {},
+                start: { line: startLine } = {}
+              } = {}
+            } = d;
+            let outputCode = fileContent;
+            let dataStart = baseLine;
+            let dataLine = `${startLine}-${endLine}`;
+            // If the code length is creater than threshold
```
Comment (boopathi) [inline client/components/Content/Content.jsx]: nit: `s/creater/greater`

File: client/components/Content/Content.jsx [hunk 36]
```diff
@@ -0,0 +1,141 @@
...
+              }
+              // Reset the starting line to the current start
+              dataStart = start;
+              // Reset the lines to highlight
+              dataLine = `${codeThreshold}-${codeThreshold +
+                (endLine - startLine)}`;
+              // Save sliced array to outputCode
+              outputCode = file.slice(start, end).join("\n");
+            }
+            const pretttPrintProps = Object.assign(
```
Comment (boopathi) [inline client/components/Content/Content.jsx]: object.assign is unnecessary

File: client/components/Sidebar/Sidebar.jsx [hunk 37]
```diff
@@ -0,0 +1,51 @@
...
+
+const BASE_CLASS = "sidebar";
+
+const Sidebar = ({ data, handleItemChange, activeIndex }) => {
+  // Get the length of the matching function
+  const getFirstFunctionMatchLength = list => {
+    if (!list.length) {
+      return null;
+    }
+    if (!list[0] || !list[0].matchingCode) {
```
Comment (boopathi) [inline client/components/Sidebar/Sidebar.jsx]: Can you give a sample case where `list[0]` is falsy ?

File: client/index.jsx (hunk 14 above)
Comment (boopathi) [reply client/index.jsx]: we are going to bundle everything and inline the script at the end of the body. There are no external scripts or stylesheets. We can control the order of loading.

File: client/webpack.config.js [hunk 38]
```diff
@@ -0,0 +1,104 @@
...
+      {
+        test: /\.css$/,
+        include: /node_modules/,
+        use: isDev
+          ? [
+              "style-loader",
+              {
+                loader: "css-loader",
+                options: {
+                  minimize: true
```
Comment (boopathi) [inline client/webpack.config.js]: minimize for dev ?

File: client/webpack.config.js [hunk 39]
```diff
@@ -0,0 +1,104 @@
...
+                }
+              }
+            ]
+          : ExtractTextPlugin.extract({
+              fallback: "style-loader",
+              use: [
+                {
+                  loader: "css-loader",
+                  options: {
+                    minimize: false
```
Comment (boopathi) [inline client/webpack.config.js]: and not for prod ?

File: client/webpack.config.js [hunk 40]
```diff
@@ -0,0 +1,104 @@
...
+      {
+        test: /\.css$/,
+        exclude: /node_modules/,
+        use: isDev
+          ? [
+              "style-loader",
+              {
+                loader: "css-loader",
+                options: {
+                  minimize: true
```
Comment (boopathi) [inline client/webpack.config.js]: same as above

File: client/webpack.config.js (hunk 38 above)
Comment (sylvesteraswin) [reply client/webpack.config.js]: By mistake! Was using this for some testing.

File: client/components/Sidebar/Sidebar.jsx (hunk 37 above)
Comment (sylvesteraswin) [reply client/components/Sidebar/Sidebar.jsx]: This is just a safety check, since we are assuming `0` index to have `matchingCode`.

Comment (boopathi) [reply client/components/Sidebar/Sidebar.jsx]: Don't assume safety check. It's better to fail and fix the bug rather than work around for not having a bug.

File: client/components/Collapse.jsx [hunk 41]
```diff
@@ -0,0 +1,17 @@
+import { h, Component } from "preact";
+import cx from "classnames";
+
+import styles from "./Collapse.css";
+
+export default class Collapse extends Component {
+  constructor(...args) {
+    super(...args);
+    this.state = {
```
Comment (vigneshshanmugam) [inline client/components/Collapse.jsx:9]: is this state even used anywhere?

File: package.json [hunk 42]
```diff
@@ -10,20 +10,11 @@
...
+    "dev": "bash -c 'cat <(find ./public/*) <(find ./lib/*) | entr -s \"./bin/js-cpa.js -m -t 1 -R html test/fixtures/*.js\"'"
   },
-  "keywords": [
-    "code",
-    "repetition",
-    "pattern",
-    "duplication",
-    "javascript"
-  ],
+  "keywords": ["code", "repetition", "pattern", "duplication", "javascript"],
```
Comment (boopathi) [inline package.json]: @vigneshshanmugam 
change your prettier config

File: lib/cli.js [hunk 43]
```diff
@@ -108,16 +130,57 @@ if (program.filelist) {
...
+      process.exit(1);
+    });
+
+  function warn(message) {
+    process.stderr.write(
+      program.colors ? chalk.yellow(chalk.bold(message)) : message
+    );
+  }
+  function write(message) {
+    process.stderr.write(program.colors ? chalk.green(message) : message);
```
Comment (vigneshshanmugam) [inline lib/cli.js:178]: shouldn't this be stdout?

File: lib/cpa.js [hunk 44]
```diff
@@ -24,6 +24,12 @@ module.exports = class CPA {
     this.threshold = threshold;
   }
 
+  // Exposing it from here for using in reporter
+  // Avoids reading from FS again
+  getSource() {
```
Comment (vigneshshanmugam) [inline lib/cpa.js:29]: added this to avoid reading from FS on html-reporter.
